Introduce dynamic hook loading

nupplaphil opened the pull request:
- Dynamically load addon files
- Dynamically load hooks
- Rewrite Logger-logic to use new hook logic (Monolog is working again)

Monolog fix: https://git.friendi.ca/friendica/friendica-addons/pulls/1400 

ToDo
- [x] Add as much phpunit-tests as possible :)
- [x] Test on friendica.philipp.info

I'm now finished. It's running on friendica.philipp.info

I tried to make parts of the logic easier in removing the "Decorator" hook logic. It seems like they're just used for edge cases, which can be covered directly in factories (Like for the `ProfilerLogger`).

I already prepared a followup for pipelining & event hook logic, which will cover all other types of hooks so far.

This hook logic can be used everywhere in the codebase to decouple namespaces .. It would be at least an entry point for "real" Domain Driven Design, which I'm no more forcing, but trying to add as much potential entrypoints as possible ;-)

[Edit] Btw. the woodpecker CI instance isn't working properly. I think I will have some time in the next 2 weeks to hopefully fix it...

nupplaphil committed:
Introduce dynamic hook loading
- Dynamically load addon files
- Dynamically load hooks
- Rewrite Logger-logic to use new hook logic (Monolog is working again)

nupplaphil committed:
Fix license check

nupplaphil committed:
Make PHP-CS happy

nupplaphil committed:
Some more PHP-CS

nupplaphil committed:
Remove unnecessary trailing slashes

nupplaphil committed:
Doc

nupplaphil committed:
Add doc

nupplaphil committed:
improve hooks.config.php

nupplaphil committed:
Update messages.po

nupplaphil committed:
add docker-translate command

nupplaphil committed:
Update woodpecker PHP version

nupplaphil committed:
Add tests for AddonFiles

nupplaphil committed:
Add tests for HookFileManager

nupplaphil committed:
Add tests for InstanceManager and remove Decorator hook logic (avoid complex Dice logic)

nupplaphil committed:
Add license

nupplaphil committed:
Make PHP-CS happy

nupplaphil committed:
Fix logger classes and tests

nupplaphil committed:
Fix SyslogLogger and tests

nupplaphil committed:
Fix PHP-CS

nupplaphil committed:
Remove unused ProfilerLogger factory

nupplaphil committed:
Restore LoggerInterface import

MrPetovan commented on bin/daemon.php:
```diff
@@ -60,7 +60,10 @@
+/** @var \Friendica\Core\Addon\Capabilities\ICanLoadAddons $addonLoader */
+$addonLoader = $dice->create(\Friendica\Core\Addon\Capabilities\ICanLoadAddons::class);
+$dice = $dice->addRules($addonLoader->getActiveAddonConfig('dependencies'));
+$dice = $dice->addRule(LoggerInterface::class,['constructParams' => [Logger\Capabilities\LogChannel::DAEMON]]);
```
```suggestion
$dice = $dice->addRule(LoggerInterface::class, ['constructParams' => [Logger\Capabilities\LogChannel::DAEMON]]);
```

MrPetovan commented on bin/worker.php:
```diff
@@ -54,7 +55,10 @@
+/** @var \Friendica\Core\Addon\Capabilities\ICanLoadAddons $addonLoader */
+$addonLoader = $dice->create(\Friendica\Core\Addon\Capabilities\ICanLoadAddons::class);
+$dice = $dice->addRules($addonLoader->getActiveAddonConfig('dependencies'));
+$dice = $dice->addRule(LoggerInterface::class,['constructParams' => [LogChannel::WORKER]]);
```
```suggestion
$dice = $dice->addRule(LoggerInterface::class, ['constructParams' => [LogChannel::WORKER]]);
```

MrPetovan commented on src/Core/Addon/Exception/AddonInvalidConfigFileException.php:
```diff
@@ -0,0 +1,35 @@
+ */
+class AddonInvalidConfigFileException extends \RuntimeException
+{
+	public function __construct($message = "", $code = 0, Throwable $previous = null)
```
```suggestion
	public function __construct($message = '', $code = 0, Throwable $previous = null)
```

MrPetovan commented on src/Core/Hooks/Capabilities/ICanCreateInstances.php:
```diff
@@ -0,0 +1,41 @@
+	 *
+	 * @return object The concrete instance of the type "$class"
+	 */
+	public function create(string $class, string $name, array $arguments = []): object;
```
```suggestion
	public function create(string $class, string $strategy, array $arguments = []): object;
```

MrPetovan commented on src/Core/Hooks/Capabilities/ICanRegisterInstances.php:
```diff
@@ -0,0 +1,46 @@
+/**
+ * Register strategies for given classes
+ */
+interface ICanRegisterInstances
```
Shouldn't it be `ICanRegisterStrategy`?

MrPetovan commented on src/Core/Hooks/Capabilities/ICanRegisterInstances.php:
```diff
@@ -0,0 +1,46 @@
+	 * @param string  $interface The interface, which the given class implements
+	 * @param string  $class     The fully-qualified given class name
+	 *                           A placeholder for dependencies is possible as well
+	 * @param ?string $name      An arbitrary identifier for the given class, which will be used for factories, dependency injections etc.
```
For the given class or the given strategy?

nupplaphil replied:
yep, you're right

MrPetovan commented on src/Core/Hooks/Exceptions/HookConfigException.php:
```diff
@@ -19,11 +19,12 @@
-interface IAmAStrategy
+class HookConfigException extends \RuntimeException
 {
+	public function __construct($message = "", \Throwable $previous = null)
```
```suggestion
	public function __construct($message = '', \Throwable $previous = null)
```

MrPetovan commented on src/Core/Hooks/Util/HookFileManager.php:
```diff
@@ -0,0 +1,107 @@
+	{
+		// In case it wasn't used before, reload the whole hook config
+		if (empty($this->hookConfig)) {
+			$this->reloadHookConfig();
```
Shouldn't this be done in the constructor?

nupplaphil replied:
moved the load-routine to the dependency file

MrPetovan commented on src/Core/Logger/Exception/LogLevelException.php:
```diff
@@ -23,6 +23,9 @@
+ */
 class LogLevelException extends \InvalidArgumentException
 {
 	public function __construct($message = "", Throwable $previous = null)
```
```suggestion
	public function __construct($message = '', Throwable $previous = null)
```

MrPetovan commented on src/Core/Logger/Exception/LoggerArgumentException.php:
```diff
@@ -23,6 +23,9 @@
+ */
 class LoggerArgumentException extends \InvalidArgumentException
 {
 	public function __construct($message = "", Throwable $previous = null)
```
```suggestion
	public function __construct($message = '', Throwable $previous = null)
```

MrPetovan commented on src/Core/Logger/Exception/LoggerException.php:
```diff
@@ -23,6 +23,9 @@
+ */
 class LoggerException extends \Exception
 {
 	public function __construct($message = "", Throwable $previous = null)
```
```suggestion
	public function __construct($message = '', Throwable $previous = null)
```

MrPetovan commented on src/Core/Logger/Exception/LoggerUnusableException.php:
```diff
@@ -23,7 +23,10 @@
+ */
+class LoggerUnusableException extends \RuntimeException
 {
 	public function __construct($message = "", Throwable $previous = null)
```
```suggestion
	public function __construct($message = '', Throwable $previous = null)
```

MrPetovan commented on src/Core/Logger/Factory/AbstractLoggerTypeFactory.php:
```diff
@@ -0,0 +1,80 @@
+				return LogLevel::INFO;
+			// legacy DATA
+			case "4":
+				// legacy ALL
```
```suggestion
			// legacy ALL
```

MrPetovan commented on src/Core/Logger/Factory/Logger.php:
```diff
@@ -22,134 +22,46 @@
-	 * @return LoggerInterface The PSR-3 compliant logger instance
-	 */
-	public function create(string $loglevel = null): LoggerInterface
+	public function create(ICanCreateInstances $createInstances, IManageConfigValues $config, Profiler $profiler): LoggerInterface
```
```suggestion
	public function create(ICanCreateInstances $instanceCreator, IManageConfigValues $config, Profiler $profiler): LoggerInterface
```

MrPetovan commented on src/DI.php:
```diff
@@ -295,6 +297,11 @@ public static function flushLogger()
 		static::init($flushDice);
 	}
 
+	public static function loggCheck(): ICheckLoggerSettings
```
```suggestion
	public static function logCheck(): ICheckLoggerSettings
```

MrPetovan reviewed (changes requested):
How do the hooks declared in this PR related to existing hooks?

nupplaphil commented:
> How do the hooks declared in this PR related to existing hooks?

The existing hooks are all different kinds. This PR introduces the strategy hooks, which are used for dynamically load instance "strategies", like for
- Cache
- Storage
- Lock
- Logger

Currently, the hook returns an array and you need to extract the instance out of an array-index, like

```php
$data = [
  'instance' => null,
  'name' => 'aName',
];

Hook::callAll('instance_creation', $data);

if ($data['instance'] !== null) {
   $instance = $data['instance'];
}
```

now, it's called:
```php
$instance = $instanceRegister->create(InstanceInterface::class, 'aName');
```

and instead of 
```php
Hook::register('instance_creation', 'aName_instance_creation()`);
```

you need to set at `{addon}/static/hooks.config.php` or `static/hooks.config.php`:
```php
  HookType::STRATEGY => [
     InstanceInterface::class => [
        ANameInstance::class => ['aName'],
     ],
  ],
```
which automatically gets registered at runtime.

nupplaphil committed:
Apply suggestions from code review

Co-authored-by: Hypolite Petovan <hypolite@mrpetovan.com>

MrPetovan commented:
I really would like to avoid the confusion between the existing data hooks and the new instance hooks. Can we just use the term "Strategies" instead of "Strategy Hooks"?

nupplaphil commented:
In the end, all hooks are replaced by either strategies, pipelines or events. But yes, this naming convention would work as well. But they should stay inside the hooks namespace, shouldn't it?

Or do you have a different common namespace in mind?

I see this pattern as a possibility for decoupling the core logic as well :-).

Events for one-way events to all listeners, pipelines for dynamically process content and strategies to choose different implementations.

But that's just a possibility, for now they should replace the `Hook` logic and corresponding table.

MrPetovan commented:
Ah, I see what you mean. In this case yes, the overarching Hook namespace still works. The transition might be a little confusing with a `hooks.config.php` that doesn't have anything to do with the current hooks.

nupplaphil commented:
Shall I rename it? I'm open to any suggestions :D

MrPetovan commented:
I'd rather we had a separate `strategies.config.php` file until we have all concepts under the same umbrella.

nupplaphil committed:
Adhere feedback
- rename hooks.config.php to strategies.config.php
- change all corresponding classes and tests

nupplaphil committed:
Make PHP-CS happy

nupplaphil commented:
I'm finished here ;-)
